hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,246 @@
+
+## Solution Proposal
+
+The proposed solution involves defining a container level checksum that can be used to quickly tell if two containers replicas match or not based on their data. This container checksum can be defined as a three level Merkle tree:
+
+- Level 1 (leaves): The existing chunk level checksums (written by the client and verified by the existing datanode container scanner).
+- Level 2: A block level checksum created by hashing all the chunk checksums within the block.
+- Level 3 (root): A container level checksum created by hashing all the block checksums within the container.
+    - This top level container hash is what is reported to SCM to detect diverged replicas. SCM does not need block or chunk level hashes.
+
+When SCM sees that replicas of a non-open container have diverged container checksums, it can trigger a reconciliation process on all datanode replicas. SCM does not need to know which container hash is correct (if any of them are correct), only that all containers match. Datanodes will use their merkle tree and those of the other replicas to identify issues with their container. Next, datanodes can read the missing data from existing replicas and use it to repair their container replica.
+
+### Phase I (outlined in this document)
+
+- Add container level checksums that datanodes can compute and store.
+
+- Add a mechanism for datanodes to reconcile their replica of a container with another datanode's replica so that both replicas can be verified to be equal at the end of the process.
+
+- Add a mechanism for SCM to trigger this reconciliation as part of the existing heartbeat command protocol SCM uses to communicate with datanodes.
+
+- Add an `ozone admin container reconcile <container-id>` CLI that can be used to manually resolve diverged container states among non-open container replicas.
+
+### Phase II (out of scope for this document)
+
+- Automate container reconciliation requests as part of SCM's replication manager.
+
+- Simplify SCM replication manager decommission and recovery logic based on mismatch of container checksums, instead of the combination of all possible container states.
+
+### Phase III (out of scope for this document)
+
+- Extend container level checksum verification to erasure coded containers.
+    - EC container replicas do not have the same data and are not expected to have matching container checksums.
+    - EC containers already use offline recovery as a reconciliation mechanism.
+
+## Solution Implementation
+
+### Storage
+
+The only extra information we will need to store is the container merkle tree on each datanode container replica. The current proposal is store this separately as a proto file on disk so that it can be copied over the network exactly as stored. The structure would look something like this (not finalized, for illustrative purposes only):
```

Comment:
We should not limit the number of blocks in a container to avoid container explosion due to small files. We can run some rough numbers to see the tradeoff between block count vs hash tree size.

Reply:
For tiny blocks, say less than 1k, we may write the block data to rocksdb (in the future) instead of storing it as a file.  This can avoid creating a lot of small files on disk.  We should think about how to optimize the checksum tree.  This may not be a big problem since container size is only 5GB.  The checksum tree size should be < 1GB.

Reply:
If any block is deleted later,  will it require recalculate the tree and update the on disk file?